compiler-core/src/language_server/code_action.rs

```diff
@@ -1043,3 +1043,566 @@ impl<'a> AddAnnotations<'a> {
+        if overlaps(self.params.range, range) {
+            if let Some((module_alias, module_location)) = module {
+                if let Some(import) =
+                    self.module
+                        .find_node(module_location.start)
+                        .and_then(|node| {
+                            if let Located::Annotation(_, ty) = node {
+                                if let Some((module, _)) = ty.named_type_name() {
+                                    return self.get_module_import(&module);
+                                }
+                            }
+                            None
+                        })
+                {
+                    self.qualified_constructor = Some(QualifiedConstructor {
+                        import,
+                        module_aliased: import.as_name.is_some(),
+                        used_name: module_alias.clone(),
+                        constructor: name.clone(),
+                        is_type: true,
+                    });
+                }
+            }
+        }
+        ast::visit::visit_type_ast_constructor(self, location, module, name, arguments);
+    }
+
+    fn visit_typed_expr_module_select(
+        &mut self,
+        location: &'ast SrcSpan,
+        type_: &'ast Arc<Type>,
+        label: &'ast EcoString,
+        module_name: &'ast EcoString,
+        module_alias: &'ast EcoString,
+        constructor: &'ast ModuleValueConstructor,
+    ) {
+        let range = src_span_to_lsp_range(*location, &self.line_numbers);
+        if overlaps(self.params.range, range) {
+            if let ModuleValueConstructor::Record {
+                name: constructor_name,
+                ..
+            } = constructor
+            {
+                if let Some(import) = self.get_module_import(module_name) {
+                    self.qualified_constructor = Some(QualifiedConstructor {
+                        import,
+                        module_aliased: import.as_name.is_some(),
+                        used_name: module_alias.clone(),
+                        constructor: constructor_name.clone(),
+                        is_type: false,
+                    });
+                }
+            }
+        }
+        ast::visit::visit_typed_expr_module_select(
+            self,
+            location,
+            type_,
+            label,
+            module_name,
+            module_alias,
+            constructor,
+        )
+    }
+
+    fn visit_typed_pattern_constructor(
+        &mut self,
+        location: &'ast SrcSpan,
+        name: &'ast EcoString,
+        arguments: &'ast Vec<CallArg<TypedPattern>>,
+        module: &'ast Option<(EcoString, SrcSpan)>,
+        constructor: &'ast crate::analyse::Inferred<type_::PatternConstructor>,
+        spread: &'ast Option<SrcSpan>,
+        type_: &'ast Arc<Type>,
+    ) {
+        let range = src_span_to_lsp_range(*location, &self.line_numbers);
+        if overlaps(self.params.range, range) {
```

Comment:
I think we could avoid the extra work if we've found the constructor already, seeing as it does another module traversal within the `if` body.
```suggestion
        if self.qualified_constructor.is_none() && overlaps(self.params.range, range) {
```

Reply:
I think for record value constructor yes, but for type constructors or pattern record constructor say `option.Option(wobble.Wobble)` the whole `option.Option(wobble.Wobble)` is valid range so when trigger at `wobble.Wobble` it would go through `option.Option` first which will set the qualified_constructor.